docs/book/v2/routing-middleware.md

```diff
@@ -0,0 +1,263 @@
+                        },
+                    ],
+                ],
+            ],
+        ],
+    ],
+];
+```
+
+## Mixed `PipeSpec` Arguments
+
+In this example multiple different possible parameters are passed to `PipeSpec`:
+
+```php
+use Application\Handler\AlbumDetailHandler;
+use Laminas\Mvc\Middleware\PipeSpec;
+use Laminas\Router\Http\Literal;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+
+return [
+    'router' => [
+        'routes' => [
+            'instance-pipespec-dev-env' => [
+                'type' => Literal::class,
+                'options' => [
+                    'route' => '/instance-pipespec-dev-env',
+                    'defaults' => [
+                        'controller' => PipeSpec::class,
+                        'middleware' => new PipeSpec(
+
+                            // id/alias of service in container
+                            'container_id_for_album_middleware',
+
+                            // anonymous class implementing a MiddlewareInterface
+                            new class implements MiddlewareInterface {
+                                public function process(
+                                    ServerRequestInterface $request,
+                                    RequestHandlerInterface $handler
+                                ): ResponseInterface {
+                                    return $handler->handle($request);
+                                }
+                            },
+
+                            // Closure
+                            static function (
+                                ServerRequestInterface $request,
+                                RequestHandlerInterface $handler
+                            ): ResponseInterface {
+                                return $handler->handle($request);
+                            },
+
+                            // class-string of an Handler which will be resolved by the container
+                            AlbumDetailHandler::class,
+                        ),
+                    ],
+                ],
+            ],
+        ],
+    ],
+];
+```
+
+Please note that if you pass instances of middleware or handlers directly, it is not compatible with config caching. You
+should use that for development only. The recommended variant is passing container managed middleware
+like `SomeRequestHandler::class`.
+
+## Defining `:middleware` Placeholder in Route Definition
+
+Although you can define a `:middleware` placeholder in your middleware route definition, we are strongly advising
+against doing so. This can be utilized maliciously to invoke middleware that was never intended to be called directly.
+Since middleware lookup uses the main application container, it can also be abused to cause instantiation of expensive
+services leading to a DoS vulnerability.
```

Comment:
Question: should we maybe try and _detect_ this scenario, and raise an exception when found?

Reply:
That would be out of scope tbh. Scanning route config could be added as part of some kind of tooling package. However, since Laminas is intentionally not opinionated it would be pretty hard to build the tool that would be smart enough to find bad practices.

Reply:
maybe we could somehow mark the feature as `@internal` and being subject to change?